hadoop-hdds/docs/content/design/storage-class.md

```diff
@@ -19,10 +19,331 @@ author: Marton Ele
+ * SCM allocateContainer logic should be changed. Right now it creates a predefined number of Open containers for each `owner` (high level applocation). It can be extended easily to create open containers for each storage-classes
+ * Having slightly more open containers doesn't make any difference (5-10 more open containers per disk vs. the 5000-10000 existing containers)
+
+# Backward compatibility
+
+ * In allocateBlock and intra-service communication we can replace replication type and replication factor with storage-class
+ * In client API replication factor and replication type will be optional
+ * Legacy type/factor definitions will be mapped to real storage classes under the hood (which defines the real replication scheme). For example `Ratis/THREE --> STANDARD`
+
+# Additional use cases
+
+In addition to the possible, replication related additional options there are two very specific use cases where we can use storage classes. Both requires more design discussion but here I quickly summarize some *possible directions* with the help of the storage class abstraction.
+
+This section demonstrates the power of the abstraction the exact implementation of these features can be (and probably will be slightly) different
+
+## Erasure coding
+
+To store cold data on less bits (less than the data * 3 replicas) we can encode the data and store some additional bits to survive replica loss. In a streaming situation (streaming write) it can be tricky as we need enough chunks to do the Reed-Solomon magic. With containers we are in a better position. After the first transition of the Open containers we can do EC encoding and that time we have enough data to encode.
+
+There are multiple options to do EC, one most simplest way is to encode Containers after the first transition:
+
+Storage class COLD:
+
+ * *First container type/parameters*: Ratis/THREE replicated containers
+ * *Transitions*: In case of any error or if the container is full, convert to closed containers
+ * *Second container type/parameters*: EC RS(6,2)
+
+With this storage class the containers can be converted to a specific EC container together with other containers (For example instead of 3 x C1, 3 x C2 and 3 x C3 containers can be converted to C1, C2, C3 + Parity 1 + Parity2).
+
+## Random-read write
+
+NFS / Fuse file system might require to support Random read/write which can be tricky as the closed containers are immutable. In case of changing any one byte in a block, the whole block should be re-created with the new data. It can have a lot of overhead especially in case of many small writes.
+
+But write is cheap with Ratis/THREE containers. Similar to any `ChunkWrite` and `PutBlock` we can implementa `UpdateChunk` call which modifies the current content of the chunk AND replicates the change with the help of Ratis.
```

Comment:
we can always create new chunks and link that to the block? the old chunk can be kept as old version or garbage collected like delete.

Reply:
Yes, this is also possible. This section shows the possibilities with the storage-class abstraction. With this abstraction we can define different behavior for the buckets which are used for random read write. The exact details can be changes